src/nvim/api/vim.c

```diff
@@ -2389,3 +2389,169 @@ Dictionary nvim_complete_set(Integer index, Dict(complete_set) *opts, Arena *are
   }
   return rv;
 }
+
+enum {
+  kStatusNone = 0x0,
+  kStatusBar = 0x1,
+  kStatusColumn = 0x2,
+  kStatusLine = 0x4,
+  kStatusTab = 0x8,
+  kStatusAll = 0x10,
+};
+
+static void redraw_status(win_T *wp, int status, bool *flush)
+{
+  if (*wp->w_p_stc != NUL && (status & (kStatusColumn|kStatusAll))) {
+    wp->w_nrwidth_line_count = 0;
+    changed_window_setting(wp);
+  }
+  win_grid_alloc(wp);
+
+  // Just hid or showing winbar for the first time, or drawing the statuscolumn.
+  // Flush later because the window itself should be redrawn.
+  if (wp->w_lines_valid == 0) {
+    *flush = true;
+  }
+
+  if (*flush && (status & (kStatusBar|kStatusLine|kStatusAll))) {
+    // When flushing later, just mark for redraw.
+    wp->w_redr_status = true;
+  } else if (status & (kStatusBar|kStatusLine)) {
+    // Just redraw the 'winbar' or 'statusline'.
+    win_check_ns_hl(wp);
+    if (status & kStatusBar) {
+      win_redr_winbar(wp);
+    }
+    if (status & kStatusLine) {
+      win_redr_status(wp);
+    }
+    win_check_ns_hl(NULL);
+  }
+}
+
+/// Instruct Nvim to redraw various elements.
+///
+/// @param opts  Optional parameters.
+///               - flush: Update the screen with pending updates, even if
+///                 Nvim is waiting for input and would normally not do so.
+///               - clear: Forcefully redraw `win`, all windows containing
+///                 `buf`, or all windows when both are omitted.
+///               - range: Buffer range to redraw, useful for decoration
+///                 providers. Tuple `[first, last]`:
+///                   - first: first line number of the range
+///                   - last: last line number of the range
+///               - cursor: Immediately update cursor position on the screen.
+///               - status: Redraws (a combination of) the following status
+///                 components (array of one or more strings):
+///                 - "bar": |'winbar'|
+///                 - "column": |'statuscolumn'|
+///                 - "line": |'statusline'|
+///                 - "tab": |'tabline'|
+///                 - "all": All of the above
+///               - win: Target a specific |window-ID| for `cursor`, `status`
+///                 and `clear` updates. Current window when 0, all windows if
+///                 omitted and applicable.
+///               - buf: Target a specific buffer number for `range`, `status`,
+///                 and `clear` updates, current buffer when 0 or omitted.
+void nvim_redraw(Dict(redraw) *opts, Error *err)
+  FUNC_API_SINCE(12)
+{
+  win_T *win = NULL;
+  buf_T *buf = NULL;
+  bool save_lz = p_lz;
+  int save_rd = RedrawingDisabled;
+  RedrawingDisabled = 0;
+  p_lz = false;
```

Comment:
I think it should be enough to do these around `update_screen()` inside its if-statement.

Reply:
`win_redr_status()` checks `!redrawing()` as well.